gh-102856: Add changes related to PEP 701 in 3.12 What's New docs (#104824)
python/cpython: Doc/whatsnew/3.12.rst

New Features
============

.. _whatsnew312-pep701:

PEP 701: Syntactic formalization of f-strings
---------------------------------------------

:pep:`701` lifts some restrictions on the usage of f-strings. Expression components
inside f-strings can now be any valid Python expression including backslashes,
unicode escaped sequences, multi-line expressions, comments and strings reusing the
same quote as the containing f-string. Let's cover these in detail:

> * Quote reuse: in Python 3.11, reusing the same quotes as the contaning f-string

Reviewer: ```suggestion
* Quote reuse: in Python 3.11, reusing the same quotes as the containing f-string
```

  raises a :exc:`SyntaxError`, forcing the user to either use other available

>   quotes (like using double quotes or triple quites if the f-string uses single
>   quites). In Python 3.12, you can now do things like this:

Reviewer: ```suggestion
  quotes (like using double quotes or triple quotes if the f-string uses single
  quotes). In Python 3.12, you can now do things like this:
```

>     >>> songs = ['Take me back to Eden', 'Alkaline', 'Ascensionism']

Reviewer 2: Very good improvement.

>     >>> f"This is the playlist: {", ".join(things)}"

Reviewer 3: ```suggestion
    >>> f"This is the playlist: {", ".join(songs)}"
```

    'This is the playlist: Take me back to Eden, Alkaline, Ascensionism

  Note that before this change there was no explicit limit in how f-strings can
  be nested, but the fact that string quotes cannot be reused inside the
  expression component of f-strings made it impossible to nest f-strings

>   arbitrarily. In fact, this is the most nested-fstring that can be written:

Reviewer: ```suggestion
  arbitrarily. In fact, this is the most nested f-string that could be written:
```

  that may be surprising or dangerous.
  See :ref:`tarfile-extraction-filter` for details.

* The output of the :func:`tokenize.tokenize` and :func:`tokenize.generate_tokens`
  functions is now changed due to the changes introduced in :pep:`701`. This

>   means that ``STRING`` tokens are not emited anymore for f-strings and the

Reviewer: ```suggestion
  means that ``STRING`` tokens are not emitted any more for f-strings and the
```

  tokens described in :pep:`701` are now produced instead: ``FSTRING_START``,

>   ``FSRING_MIDDLE`` and ``FSTRING_END`` are now emited for f-string "string"

Reviewer: ```suggestion
  ``FSRING_MIDDLE`` and ``FSTRING_END`` are now emitted for f-string "string"
```

>   parts in addition to the the apropiate tokens for the tokenization in the

Reviewer: ```suggestion
  parts in addition to the appropriate tokens for the tokenization in the
```

  expression components. For example for the f-string ``f"start {1+1} end"``
  the old version of the tokenizer emitted::

    1,0-1,18:           STRING         'f"start {1+1} end"'

  while the new version emits::

      1,0-1,2:            FSTRING_START  'f"'
      1,2-1,8:            FSTRING_MIDDLE 'start '
      1,8-1,9:            OP             '{'
      1,9-1,10:           NUMBER         '1'
      1,10-1,11:          OP             '+'
      1,11-1,12:          NUMBER         '1'
      1,12-1,13:          OP             '}'
      1,13-1,17:          FSTRING_MIDDLE ' end'
      1,17-1,18:          FSTRING_END    '"'

>   Aditionally, final ``DEDENT`` tokens are now emited within the bounds of the

Reviewer: ```suggestion
  Aditionally, final ``DEDENT`` tokens are now emitted within the bounds of the
```
